betty/plugin/__init__.py

```diff
@@ -0,0 +1,177 @@
+"""
+The Plugin API.
+
+Plugins allow third-party code (e.g. your own Python package) to add functionality
+to Betty.
+
+Read more at :doc:`/usage/plugin`.
+"""
+
+from __future__ import annotations
+
+import re
+from typing import TypeVar, Generic, Self, overload, TYPE_CHECKING, TypeAlias, TypeGuard
+
+if TYPE_CHECKING:
+    from betty.locale import Localizable
+    from collections.abc import AsyncIterator, Sequence
+
+
+PluginId: TypeAlias = str
+"""
+A plugin ID is a string that meets these criteria:
+- At most 255 characters long.
+- Lowercase letters, numbers, hyphens (-), and underscores (_) only.
+
+See :py:func:`betty.plugin.validate_plugin_id`.
+"""
+
+
+PLUGIN_ID_PATTERN = re.compile(r"^[a-z0-9\-_]{1,250}$")
```

Comment:
Make this private. The validation API method is the only thing that should be public.